RXPatchLib/UpdateServerHandler.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace RXPatchLib
+{
+
+    /// <summary>
+    /// Contains a server entry that eventually sits inside a List array
+    /// </summary>
+    public class UpdateServerEntry
+    {
+        public Uri Uri;
+        public string FriendlyName;
+        public bool IsUsed;
+        public bool HasErrored;
+        public string WebPatchPath;
+        public long Latency;
+
+        public UpdateServerEntry(string Url, string FriendlyName)
+        {
+            Uri = new Uri(Url);
+            this.FriendlyName = FriendlyName;
+        }
+    }
+
+    /// <summary>
+    /// Handles all Update Server URLs and Friendly Names, including getting the best latency server
+    /// </summary>
+    internal class UpdateServerHandler
+    {
+        private List<UpdateServerEntry> _updateServers = new List<UpdateServerEntry>();
+
+        public void AddUpdateServer(string Url, string FriendlyName)
+        {
+            try
+            {
+                _updateServers.Add(new UpdateServerEntry(Url, FriendlyName));
+            }
+            catch
+            {
+                // ignored
+            }
+        }
+
+        public List<UpdateServerEntry> GetUpdateServers()
+        {
+            return _updateServers;
+        }
+
+        /// <summary>
+        /// Selects the best patch server in the list that both is not in use, and has not errored
+        /// </summary>
+        /// <returns>An UpdateServerEntry of the host found, or Null if no more hosts exist</returns>
+        public UpdateServerEntry SelectBestPatchServer()
+        {
+            // Find out the best latency server to respond with
+            foreach (var entry in _updateServers.Where(x => x.Latency == 0 && !x.IsUsed && !x.HasErrored))
+            {
+                GetServerLatency(entry);
+            }
+
+            _updateServers = _updateServers.OrderBy(x => x.Latency).ToList();
+
+            return _updateServers.DefaultIfEmpty(null).FirstOrDefault(x => !x.HasErrored && !x.IsUsed);
+        }
+
+        /// <summary>
+        /// Tests the latency of the remote server for the user
+        /// </summary>
+        /// <param name="entry">The UpdateServerEntry object to test against</param>
+        /// <returns>Returns the latency as a float</returns>
+        private void GetServerLatency(UpdateServerEntry entry)
```

Comment:
It seems like the key difference between the previous algorithm and the current one is here, where you've changed it to determine mirror reliability sequentially rather than in parallel. While this should be more reliable in determining which mirror is the best, I'm concerned that it will potentially greatly increase the time required to determine the "best" mirror. For example since the timeout is currently set to 1500ms, if we have 11 mirrors and 10 are inaccessible by the user (for whatever reason), it will take more than 15 seconds before we can start downloading. The current mirror selection algorithm on the other hand allows downloading to proceed as soon as the first mirror responds (which is presumably the "best" mirror since it was the fastest to respond). Additionally there's a minor concern that this may be inaccurate if the user's network traffic fluctuates significantly during the mirror selection process.

All of this considered I think we should avoid any significant modification of the current algorithm, and instead move towards concurrent downloads such that we can mitigate the importance of the mirror selection algorithm or even remove it entirely by using all available download mirrors at all times. Issue #12 currently indicates that we'll use only 4 simultaneous mirrors, but this could (and perhaps should) be subject to change. We shouldn't focus too much on server selection at this point, since it's very possible/likely to be removed entirely in the future.